futures-core/src/stream_obj.rs

```diff
@@ -0,0 +1,161 @@
+//! Asynchronous stream object value
+
+use std::fmt;
+use std::marker::{PhantomData,Unpin};
+use std::task::{ Context, Poll };
+use core::mem::PinMut;
+
+use crate::stream::Stream;
+
+/// Stealing code from core::future::future_obj
+pub struct LocalStreamObj<'a, T> {
+    ptr: *mut (),
+    poll_next_fn: unsafe fn(*mut (), &mut Context) -> Poll<Option<T>>,
+    drop_fn: unsafe fn(*mut ()),
+    _marker: PhantomData<&'a ()>,
+}
+
+impl<'a, T> Unpin for LocalStreamObj<'a, T> {}
+
+impl<'a, T> LocalStreamObj<'a, T> {
+    /// Create a `LocalFutureObj` from a custom trait object representation.
+    #[inline]
+    pub fn new<F: UnsafeStreamObj<'a, T> + 'a>(f: F) -> LocalStreamObj<'a, T> {
+        LocalStreamObj {
+            ptr: f.into_raw(),
+            poll_next_fn: F::poll_next,
+            drop_fn: F::drop,
+            _marker: PhantomData,
+        }
+    }
+
+    /// Converts the `LocalStreamObj` into a `StreamObj`
+    /// To make this operation safe one has to ensure that the `UnsafeStreamObj`
+    /// instance from which this `LocalStreamObj` was created actually
+    /// implements `Send`.
+    #[inline]
+    pub unsafe fn into_stream_obj(self) -> StreamObj<'a, T> {
+        StreamObj(self)
+    }
+}
+
+impl<'a, T> fmt::Debug for LocalStreamObj<'a, T> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_struct("LocalFutureObj")
+            .finish()
+    }
+}
+
+impl<'a, T> From<StreamObj<'a, T>> for LocalStreamObj<'a, T> {
+    #[inline]
+    fn from(f: StreamObj<'a, T>) -> LocalStreamObj<'a, T> {
+        f.0
+    }
+}
+
+impl<'a, T> Stream for LocalStreamObj<'a, T> {
+    type Item = T;
+
+    #[inline]
+    fn poll_next(self: PinMut<Self>, cx: &mut Context) -> Poll<Option<T>> {
+        unsafe {
+            (self.poll_next_fn)(self.ptr, cx)
+        }
+    }
+}
+
+impl<'a, T> Drop for LocalStreamObj<'a, T> {
+    fn drop(&mut self) {
+        unsafe {
+            (self.drop_fn)(self.ptr)
+        }
+    }
+}
+
+/// A custom trait object for polling futures, roughly akin to
+/// `Box<dyn Future<Output = T> + Send + 'a>`.
```

Comment:
Future ^^'

Reply:
You should do a quick case-insensitive text search for "future" to find all these